File: src/Solver.cpp
#include "../include/Solver.h"


std::vector<int> select_random_genes(std::vector<Range> ranges, RngStream *prng)
{
    // randomly choose a gene from within each set of genes (represented by intervals)
    std::vector<int> result = std::vector<int>(ranges.size(), 0);
    int low, high;
    for (size_t index = 0; index != ranges.size(); ++ index)
    {
        low = ranges[index].low;
        high = ranges[index].high;
        result[index] = prng->RandInt(low, high-1);
    }

    return result;
}


void assign_pairwise_distance_score(Solution& sol, std::vector< std::vector<double> > distance_matrix)
{
    // assign score to solution, sum of all pairwise distances
    double score = 0.;
    std::vector<int> genes = sol.get_genes();
    int len = genes.size();
    int gene1, gene2;
    for (int i = 0; i < len; ++i)
    {
        gene1 = genes[i];
        for (int j = 0; j < len; ++j)
        {
            // add the distance
            gene2 = genes[j];
            score += distance_matrix[gene1][gene2];
        }
    }
    sol.set_score(score);
    return;
}


int_double find_best_substitution(std::vector<int> genes,
                                  int replacement_position,
                                  std::vector< std::vector<double> > distance_matrix,
                                  std::vector<Range> ranges)
{
    // find the best replacement for the allele at position
    double original_cost, current_cost, best_cost;
    int original_gene, best_gene;
    int gene_length = genes.size();

    // calculate the original cost (without change)
    original_cost = 0.;
    original_gene = genes[replacement_position];
    for (int i = 0; i != gene_length; ++i)
    {
        original_cost += distance_matrix[genes[i]][original_gene];
    }

    // for every possible substitution of the gene at replacement position
    // check if improvement is possible and remember the best
    best_gene = genes[replacement_position];
    best_cost = std::numeric_limits<double>::max();

    for (int current_gene = ranges[replacement_position].low;
            current_gene != ranges[replacement_position].high;
            ++current_gene)
    {

        // calculate the cost with the current replacement
        genes[replacement_position] = current_gene;
        current_cost = 0.;
        for(int i = 0; i != gene_length; ++i)
        {
            current_cost += distance_matrix[genes[i]][current_gene];
        }

        // verify if best found yet, if so, remember the new gene
        if (current_cost <= best_cost)
        {
            best_gene = current_gene;
            best_cost = current_cost;
        }
    }
    int_double ret = {best_gene, best_cost - original_cost};
    return ret;
}


void steepest_improvement(Solution& sol,
                          std::vector< std::vector<double> > distance_matrix,
                          std::vector<Range> ranges,
                          int max_iter)
{
    // steepest descent by finding best substitution of a gene
    assert(max_iter > 0);

    // some declarations
    int best_position, best_substitution;
    int length = sol.get_genes().size();
    double best_score;
    int_double substitution;
    int iteration = 0;

    while (iteration < max_iter)
    {
        iteration += 1;
        best_position = -1;
        best_substitution = -1;
        best_score = 0.;
        for (int replacement_position = 0; replacement_position != length; ++replacement_position)
        {
            substitution = find_best_substitution(sol.get_genes(), replacement_position, distance_matrix, ranges);
            if (substitution.cost_difference < best_score)
            {
                best_position = replacement_position;
                best_substitution = substitution.best_gene;
                best_score = substitution.cost_difference;
            }
        }
        // check if a local minima is reached
        if (best_substitution == -1)
        {
            break;
        }
        else
        {
            sol.set_gene(best_position, best_substitution);
        }
    }

    // recalculate the score
    assign_pairwise_distance_score(sol, distance_matrix);
    return;
}



std::vector<Solution> initialize_population(std::vector<Range> ranges,
                                            int pop_size,
                                            RngStream *prng)
{
    // initialize the population
    assert(pop_size > 0);
    std::vector<Solution> population = std::vector<Solution>();

    // create random initial, until the population is filled
    for (int iteration = 0; iteration != pop_size; ++iteration)
    {
        population.push_back(Solution(select_random_genes(ranges, prng)));
    }
    return population;
}

File: java/src/Solver.java
import java.util.ArrayList;

public abstract class Solver {


    // static methods
    /**
     * calculates and assigns the sum of pairwise distances as the new score of the given solution
     * @param solution solution to evaluate
     * @param distanceMatrix pre-calculated distance matrix
     */
    public static void assignPairwiseDistanceScore(Solution solution, double[][] distanceMatrix)
    {
        double score = 0.;
        ArrayList<Integer> genes = solution.getGenes();
        for (int gene1 : genes)
        {
            for (int gene2 : genes)
            {
                score += distanceMatrix[gene1][gene2];
            }
        }
        solution.setScore(score);
    }


    /**
     * Randomly select one gene per interval (range) and return a list of it
     * @param ranges intervals from which to select from
     * @param stream pseudo-random number generator stream (L'Ecuyer)
     * @return list, one gene per interval
     */
    private static ArrayList<Integer> selectRandomGenes(ArrayList<Pair <Integer, Integer>> ranges,
                                                       RngStream stream)
    {
        ArrayList<Integer> genes = new ArrayList<>();
        for (Pair<Integer, Integer> range : ranges)
        {
            genes.add(stream.randInt(range.getFirst(), range.getSecond() - 1));
        }
        return genes;
    }


    /**
     * Using the distance matrix, find the change of gene that brings the most improvement (greedy)
     * @param genes current genes chosen
     * @param replacement_position index of the gene list to investigate
     * @param distance_matrix pre-calculated matrix of distances
     * @param ranges list of intervals from which to select new genes from
     * @return index of the new gene (-1 if no better) and difference in score
     */
    public static Pair<Integer, Double> findBestSubstitution(ArrayList<Integer> genes,
                                                             int replacement_position,
                                                             double[][] distance_matrix,
                                                             ArrayList<Pair<Integer, Integer>> ranges)
    {
        // find the best replacement for the allele at position
        double original_cost, current_cost, best_cost;
        int original_gene, best_gene;
        int gene_length = genes.size();

        // calculate the original cost (without change)
        original_cost = 0.;
        original_gene = genes.get(replacement_position);
        for (int i = 0; i != gene_length; ++i)
        {
            original_cost += distance_matrix[genes.get(i)][original_gene];
        }

        // for every possible substitution of the gene at replacement position
        // check if improvement is possible and remember the best
        best_gene = genes.get(replacement_position);
        best_cost = Double.POSITIVE_INFINITY;

        int begin = ranges.get(replacement_position).getFirst();
        int end = ranges.get(replacement_position).getSecond();
        for (int current_gene = begin; current_gene != end; ++current_gene)
        {
            // calculate the cost with the current replacement
            genes.set(replacement_position, current_gene);
            current_cost = 0.;
            for(int i = 0; i != gene_length; ++i)
            {
                current_cost += distance_matrix[genes.get(i)][current_gene];
            }

            // verify if best found yet, if so, remember the new gene
            if (current_cost <= best_cost)
            {
                best_gene = current_gene;
                best_cost = current_cost;
            }
        }
        return new Pair<>(best_gene, best_cost - original_cost);
    }




    /**
     * applies the steepest descent procedure until either
     * a local minima is found or maxNumIterations iterations of the substitution are done
     * @param solution solution to improve
     * @param distanceMatrix pre-calculated matrix of distances
     * @param ranges list of intervals from which to select from
     * @param maxNumIterations maximum number of iterations to perform steepest descent
     */
    public static void steepestDescent(Solution solution,
                                       double[][] distanceMatrix,
                                       ArrayList<Pair<Integer, Integer>> ranges,
                                       int maxNumIterations)
    {
        // forward declarations
        int best_position, best_substitution;
        double best_score;
        Pair<Integer, Double> substitution;
        int iteration = 0;

        while (iteration < maxNumIterations)
        {
            iteration += 1;
            best_position = -1;
            best_substitution = -1;
            best_score = 0.;
            for (int replacement_position = 0; replacement_position != solution.getGenes().size(); ++replacement_position)
            {
                substitution = findBestSubstitution(solution.getGenes(), replacement_position, distanceMatrix, ranges);
                if (substitution.getSecond() < best_score)
                {
                    best_position = replacement_position;
                    best_substitution = substitution.getFirst();
                    best_score = substitution.getSecond();
                }
            }
            // check if a local minima is reached
            if (best_substitution == -1)
            {
                break;
            }
            else
            {
                solution.setGene(best_position, best_substitution);
            }
        }

        // recalculate the score
        assignPairwiseDistanceScore(solution, distanceMatrix);
    }


    /**
     * create a list of Solutions generated randomly
     * @param ranges list of (begin, end) indices, for each gene choices
     * @param size size of the output list
     * @param stream pseudo random number generator stream
     * @return list of Solution randomly generated
     */
    public static ArrayList<Solution> initializeRandomSolutions(ArrayList<Pair<Integer, Integer>> ranges, int size, RngStream stream)
    {
        {
            // initialize the population
            assert(size > 0);
            ArrayList<Solution> solutions = new ArrayList<>();

            // create random initial, until the population is filled
            for (int iteration = 0; iteration != size; ++iteration)
            {
                solutions.add(new Solution(selectRandomGenes(ranges, stream), 0.));
            }
            return solutions;
        }
    }


    // abstract methods
    /**
     * solver interface for the consensus problem
     * @param distanceMatrix pre-calculated matrix of distances
     * @param ranges ranges of the problem
     * @return list of the best(s) solution(s)
     */
    public abstract ArrayList<Solution> solve(double[][] distanceMatrix, ArrayList<Pair<Integer, Integer>> ranges);

    /**
     * gets the verbosity setting of the solver
     * @return verbosity (yes or no)
     */
    public abstract boolean isVerbose();
}
